File: tests/test_equilibrium.py
from pathlib import Path

import pytest
import torch
from kintera import (
    EquilibriumOptions,
    EquilibriumTP,
    atomic_mass,
    molar_mass,
)
from kintera.equilibrium import Nasa9LogK


def test_python_core_binding_is_functional():
    options = (
        EquilibriumOptions()
        .components(["A", "B"])
        .phases(["gas"])
        .phase_ids([0, 0])
        .reactions(["A <=> B"])
        .gas_phase(0)
    )
    solver = EquilibriumTP(options)
    moles = torch.tensor([0.8, 0.2], dtype=torch.float64)
    result, gain, diagnostics = solver(
        torch.tensor(1000.0, dtype=torch.float64),
        torch.tensor(1.0e5, dtype=torch.float64),
        moles,
        torch.zeros(1, dtype=torch.float64),
    )
    torch.testing.assert_close(
        result, torch.tensor([0.5, 0.5], dtype=torch.float64), atol=1e-7, rtol=1e-7
    )
    torch.testing.assert_close(moles, torch.tensor([0.8, 0.2], dtype=torch.float64))
    assert gain.shape == (1, 1)
    assert diagnostics[0] == 0


def test_standalone_molar_mass_utilities():
    assert atomic_mass("H") == pytest.approx(1.008e-3)
    assert molar_mass({"H": 2.0, "O": 1.0}) == pytest.approx(18.015e-3)


def test_nasa9_yaml_equilibrium_case():
    config = Path(__file__).parents[1] / "examples" / "equilibrium_nasa9.yaml"
    options = EquilibriumOptions.from_yaml(str(config))
    solver = EquilibriumTP(options)
    model = Nasa9LogK(options)
    temp = torch.tensor(4000.0, dtype=torch.float64)
    pressure = torch.tensor(1.0e5, dtype=torch.float64)
    initial = torch.tensor([0.6, 0.2, 0.2], dtype=torch.float64)

    log_k = model(temp, pressure)
    result, _, diagnostics = solver(temp, pressure, initial, log_k)

    assert log_k.shape == (1,)
    assert torch.isfinite(log_k).all()
    assert diagnostics[0] == 0
    assert diagnostics[2] < 1.0e-7
    torch.testing.assert_close(
        result @ torch.tensor([2.0, 0.0, 2.0], dtype=result.dtype),
        initial @ torch.tensor([2.0, 0.0, 2.0], dtype=initial.dtype),
    )
    torch.testing.assert_close(
        result @ torch.tensor([0.0, 2.0, 1.0], dtype=result.dtype),
        initial @ torch.tensor([0.0, 2.0, 1.0], dtype=initial.dtype),
    )


def test_ill_conditioned_multiphase_newton_step():
    """Regression for normal-equation precision loss in a Young22 state."""
    components = [
        "MgO_silicate", "SiO2_silicate", "MgSiO3_silicate",
        "FeO_silicate", "FeSiO3_silicate", "Na2O_silicate",
        "Na2SiO3_silicate", "H2_silicate", "H2O_silicate",
        "CO_silicate", "CO2_silicate", "Fe_metal", "Si_metal",
        "O_metal", "H_metal", "H2_gas", "CO_gas", "CO2_gas",
        "CH4_gas", "O2_gas", "H2O_gas", "Fe_gas", "Mg_gas",
        "SiO_gas", "Na_gas",
    ]
    reactions = [
        "Na2SiO3_silicate <=> Na2O_silicate + SiO2_silicate",
        "0.5 SiO2_silicate + Fe_metal <=> FeO_silicate + 0.5 Si_metal",
        "MgSiO3_silicate <=> MgO_silicate + SiO2_silicate",
        "O_metal + 0.5 Si_metal <=> 0.5 SiO2_silicate",
        "2 H_metal <=> H2_silicate",
        "FeSiO3_silicate <=> FeO_silicate + SiO2_silicate",
        "2 H2O_silicate + Si_metal <=> SiO2_silicate + 2 H2_silicate",
        "CO_gas + 0.5 O2_gas <=> CO2_gas",
        "CH4_gas + 0.5 O2_gas <=> 2 H2_gas + CO_gas",
        "H2_gas + 0.5 O2_gas <=> H2O_gas",
        "FeO_silicate <=> Fe_gas + 0.5 O2_gas",
        "MgO_silicate <=> Mg_gas + 0.5 O2_gas",
        "SiO2_silicate <=> SiO_gas + 0.5 O2_gas",
        "Na2O_silicate <=> 2 Na_gas + 0.5 O2_gas",
        "H2_gas <=> H2_silicate",
        "H2O_gas <=> H2O_silicate",
        "CO_gas <=> CO_silicate",
        "CO2_gas <=> CO2_silicate",
    ]
    options = (
        EquilibriumOptions()
        .components(components)
        .phases(["silicate", "metal", "gas"])
        .phase_ids([0] * 11 + [1] * 4 + [2] * 10)
        .reactions(reactions)
        .gas_phase(2)
        .standard_pressure(1.0e5)
        .max_iter(1000)
        .ftol(1.0e-8)
        .mole_floor(1.0e-40)
    )
    initial = torch.tensor(
        [
            229.45501199233973, 7.170469124760617, 6604.002063904528,
            3.5852345623803084, 250.96641936662164, 50.19328387332432,
            7.170469124760617, 7.1704691247606165e-9,
            7.1704691247606165e-9, 7.1704691247606165e-9,
            7.1704691247606165e-9, 4476.516196031982,
            4.476516196031981e-9, 4.476516196031981e-9,
            4.476516196031981e-9, 19570.252531955153, 19.589842374329482,
            1.9589842374329483e-8, 1.9589842374329483e-8,
            1.9589842374329483e-8, 1.9589842374329483e-8,
            1.9589842374329483e-8, 1.9589842374329483e-8,
            1.9589842374329483e-8, 1.9589842374329483e-8,
        ],
        dtype=torch.float64,
    )
    log_k = torch.tensor(
        [
            -4.034640768501788, -7.62490310519855, -2.039313234293211,
            7.218657926794565, 0.7959817637548838, -0.1371317344272009,
            26.01075006664782, -4.50470154179624, 38.86385560946831,
            -3.708764843086439, 4.050968036542196, 1.772570237146461,
            5.579140721827175, 14.1841721338006, -13.701917787416148,
            -14.27932134370162, -16.21258232342021, -14.87214162753718,
        ],
        dtype=torch.float64,
    )
    result, _, diagnostics = EquilibriumTP(options)(
        torch.tensor(4500.0, dtype=torch.float64),
        torch.tensor(12599210499.047432, dtype=torch.float64),
        initial,
        log_k,
    )

    assert diagnostics[0] == 0
    assert diagnostics[2] < 1.0e-8
    assert torch.isfinite(result).all()
    assert torch.all(result > 0.0)

File: tests/test_equilibrium.cpp
#include <gtest/gtest.h>

#include <filesystem>
#include <fstream>
#include <future>
#include <kintera/equilibrium/equilibrium.hpp>
#include <kintera/species.hpp>
#include <kintera/utils/molar_mass.hpp>

#define DEVICE_TESTING_SKIP_DEFAULT_INSTANTIATION
#include "device_testing.hpp"

using namespace kintera;

namespace {

EquilibriumOptions make_options() {
  auto op = EquilibriumOptionsImpl::create();
  op->components({"A", "B"})
      .phases({"gas"})
      .phase_ids({0, 0})
      .reactions({"A <=> B"})
      .gas_phase(0)
      .standard_pressure(1.e5)
      .max_iter(50)
      .ftol(1.e-7)
      .mole_floor(1.e-20);
  return op;
}

class EquilibriumDeviceTest : public DeviceTest {};
GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(DeviceTest);

TEST_P(EquilibriumDeviceTest, IdealGasReaction) {
  if (device.type() == torch::kMPS) GTEST_SKIP();
  auto op = make_options();
  EquilibriumTP eq(op);
  eq->to(device, dtype);

  auto tensor_options = torch::device(device).dtype(dtype);
  auto temp = torch::tensor({1000., 1500.}, tensor_options);
  auto pres = torch::tensor(1.e5, tensor_options);
  auto moles = torch::tensor({{0.8, 0.2}}, tensor_options);
  auto log_k = torch::zeros({2, 1}, tensor_options);

  auto [result, gain, diag] = eq->forward(temp, pres, moles, log_k);
  EXPECT_TRUE(torch::allclose(result.select(-1, 0),
                              torch::full({2}, .5, tensor_options), 1.e-5,
                              1.e-6));
  EXPECT_TRUE(torch::allclose(result.sum(-1), torch::ones({2}, tensor_options),
                              1.e-6, 1.e-6));
  EXPECT_EQ(gain.sizes(), torch::IntArrayRef({2, 1, 1}));
  EXPECT_TRUE(torch::all(diag.select(-1, 0) == 0).item<bool>());
  EXPECT_LT(diag.select(-1, 2).max().item<double>(), 1.e-5);
  EXPECT_TRUE(
      torch::allclose(moles, torch::tensor({{0.8, 0.2}}, tensor_options)));
}

TEST_P(EquilibriumDeviceTest, EvaluatesLogReactionConstant) {
  if (device.type() == torch::kMPS) GTEST_SKIP();
  auto op = make_options();
  op->A({0.1}).B2({0.4}).B1({0.4}).C({0.3}).D1({0.1}).D2({0.05});
  EquilibriumTP eq(op);
  eq->to(device, dtype);

  auto tensor_options = torch::device(device).dtype(dtype);
  auto temp = torch::tensor({2., 4.}, tensor_options);
  auto pres = torch::tensor(1.e5, tensor_options);
  auto moles = torch::tensor({0.8, 0.2}, tensor_options);

  auto [result, gain, diag] = eq->forward(temp, pres, moles);
  auto expected_log_k = 0.1 + 0.4 / temp.square() + 0.4 / temp +
                        0.3 * temp.log() + 0.1 * temp + 0.05 * temp.square();
  auto expected_b = torch::sigmoid(expected_log_k);
  EXPECT_TRUE(torch::allclose(result.select(-1, 1), expected_b, 2.e-5, 1.e-6));
  EXPECT_TRUE(torch::all(diag.select(-1, 0) == 0).item<bool>());

  auto explicit_log_k = torch::zeros({2, 1}, tensor_options);
  auto [overridden, override_gain, override_diag] =
      eq->forward(temp, pres, moles, explicit_log_k);
  EXPECT_TRUE(torch::allclose(overridden.select(-1, 1),
                              torch::full({2}, .5, tensor_options), 1.e-5,
                              1.e-6));
}

TEST(EquilibriumOptions, RequiresCoefficientsWhenLogKIsOmitted) {
  auto op = make_options();
  EquilibriumTP equilibrium(op);
  auto tensor_options = torch::TensorOptions().dtype(torch::kFloat64);
  EXPECT_THROW(equilibrium->forward(torch::tensor(1000., tensor_options),
                                    torch::tensor(1.e5, tensor_options),
                                    torch::tensor({0.8, 0.2}, tensor_options)),
               c10::Error);
}

TEST(EquilibriumOptions, RejectsPartialLogReactionConstants) {
  auto op = EquilibriumOptionsImpl::create();
  op->components({"A", "B", "C"})
      .phases({"gas"})
      .phase_ids({0, 0, 0})
      .reactions({"A <=> B", "B <=> C"})
      .A({1.})
      .B2({0.})
      .B1({0.})
      .C({0.})
      .D1({0.})
      .D2({0.});
  EXPECT_THROW(op->validate(), c10::Error);
}

TEST(MolarMass, StandaloneUtilities) {
  EXPECT_NEAR(atomic_mass("H"), 1.008e-3, 1.e-8);
  EXPECT_NEAR(molar_mass({{"H", 2.}, {"O", 1.}}), 18.015e-3, 1.e-8);
}

TEST(Nasa9, ConcurrentFirstDatabaseAccess) {
  std::vector<std::future<torch::Tensor>> futures;
  for (int i = 0; i < 8; ++i) {
    futures.push_back(std::async(std::launch::async, [] {
      return nasa9_gibbs_rt(torch::tensor(1500., torch::kFloat64),
                            {"H2", "O2", "H2O"});
    }));
  }
  auto expected = futures.front().get();
  EXPECT_EQ(expected.sizes(), torch::IntArrayRef({3}));
  EXPECT_TRUE(torch::isfinite(expected).all().item<bool>());
  for (size_t i = 1; i < futures.size(); ++i) {
    EXPECT_TRUE(torch::allclose(futures[i].get(), expected));
  }
}

TEST(EquilibriumOptions, ReadsSpeciesAndReactionsFromYaml) {
  auto path = std::filesystem::temp_directory_path() /
              "kintera_equilibrium_options_test.yaml";
  {
    std::ofstream yaml(path);
    yaml << "phases:\n"
         << "  - name: gas\n"
         << "    thermo: ideal-gas\n"
         << "    species: [A, B]\n"
         << "species:\n"
         << "  - {name: A, composition: {H: 1}}\n"
         << "  - {name: B, composition: {H: 1}}\n"
         << "reactions:\n"
         << "  - type: equilibrium\n"
         << "    equation: 'A <=> B'\n"
         << "    log-reaction-constant: {A: 1, B1: 2}\n"
         << "  - {type: arrhenius, equation: 'B => A'}\n"
         << "equilibrium: {standard-pressure: 200000, max-iter: 12, "
            "ftol: 1.e-7}\n";
  }

  auto op = EquilibriumOptionsImpl::from_yaml(path.string());
  EXPECT_EQ(op->components(), std::vector<std::string>({"A", "B"}));
  EXPECT_EQ(op->reactions(), std::vector<std::string>({"A <=> B"}));
  EXPECT_EQ(op->phase_ids(), std::vector<int>({0, 0}));
  EXPECT_EQ(op->A(), std::vector<double>({1.}));
  EXPECT_EQ(op->B1(), std::vector<double>({2.}));
  EXPECT_EQ(op->B2(), std::vector<double>({0.}));
  EXPECT_EQ(op->C(), std::vector<double>({0.}));
  EXPECT_EQ(op->D1(), std::vector<double>({0.}));
  EXPECT_EQ(op->D2(), std::vector<double>({0.}));
  EXPECT_DOUBLE_EQ(op->standard_pressure(), 2.e5);
  EXPECT_EQ(op->max_iter(), 12);
  EquilibriumTP equilibrium(op);
  EXPECT_DOUBLE_EQ(equilibrium->stoich[0][0].item<double>(), -1.);
  EXPECT_DOUBLE_EQ(equilibrium->stoich[1][0].item<double>(), 1.);
  auto masses = molar_masses_from_yaml(path.string());
  ASSERT_EQ(masses.size(), 2u);
  EXPECT_NEAR(masses[0], 1.008e-3, 1.e-8);
  EXPECT_NEAR(masses[1], 1.008e-3, 1.e-8);
  std::filesystem::remove(path);
}

TEST(EquilibriumOptions, RejectsUnbalancedYamlReaction) {
  auto path = std::filesystem::temp_directory_path() /
              "kintera_unbalanced_equilibrium_test.yaml";
  {
    std::ofstream yaml(path);
    yaml << "phases:\n"
         << "  - {name: gas, thermo: ideal-gas, species: [A, B]}\n"
         << "species:\n"
         << "  - {name: A, composition: {H: 1}}\n"
         << "  - {name: B, composition: {H: 2}}\n"
         << "reactions:\n"
         << "  - {type: equilibrium, equation: 'A <=> B'}\n";
  }
  EXPECT_THROW(EquilibriumOptionsImpl::from_yaml(path.string()), c10::Error);
  std::filesystem::remove(path);
}

INSTANTIATE_TEST_SUITE_P(
    DeviceAndDtype, EquilibriumDeviceTest,
    testing::Values(Parameters{torch::kCPU, torch::kFloat32},
                    Parameters{torch::kCPU, torch::kFloat64},
                    Parameters{torch::kCUDA, torch::kFloat32},
                    Parameters{torch::kCUDA, torch::kFloat64}),
    [](const testing::TestParamInfo<EquilibriumDeviceTest::ParamType>& info) {
      std::string name = torch::Device(info.param.device_type).str();
      name += "_";
      name += torch::toString(info.param.dtype);
      std::replace(name.begin(), name.end(), '.', '_');
      return name;
    });

}  // namespace
